### PR TITLE
second PR first commit

### PR DESCRIPTION
**Pull request by jacobperia**

### CODE REVIEW

File: app/views/layouts/application.html.haml [hunk 1]
```diff
@@ -1,18 +1,17 @@
-!!!
```
Comment (danielpaul) [inline app/views/layouts/application.html.haml]: This is the doctype of the document. So this line would be required. https://www.w3schools.com/tags/tag_doctype.asp

File: app/views/layouts/application.html.haml [hunk 2]
```diff
@@ -1,18 +1,17 @@
...
-      = stylesheet_link_tag 'https://stackpath.bootstrapcdn.com/bootstrap/4.5.2/css/bootstrap.min.css'
-      = stylesheet_link_tag 'application', media: 'all', 'data-turbolinks-track': 'reload'
-      = javascript_pack_tag 'application', 'data-turbolinks-track': 'reload'
-    = render 'shared/navbar'
-    %p.notice= notice
-    %p.alert= alert
-    = yield
+    .container
+      %p
+        = stylesheet_link_tag 'https://stackpath.bootstrapcdn.com/bootstrap/4.5.2/css/bootstrap.min.css'
```
Comment (danielpaul) [inline app/views/layouts/application.html.haml]: stylesheets are always in the `<head>` section in the HTML

File: app/views/layouts/application.html.haml [hunk 3]
```diff
@@ -1,18 +1,17 @@
-!!!
 %html
   %head
+    = render 'shared/navbar'
```
Comment (danielpaul) [inline app/views/layouts/application.html.haml]: is this mean to be in the head section? All visible HTML should be in the body tag.

File: app/views/layouts/application.html.haml [hunk 4]
```diff
@@ -1,18 +1,17 @@
...
-      = javascript_pack_tag 'application', 'data-turbolinks-track': 'reload'
-    = render 'shared/navbar'
-    %p.notice= notice
-    %p.alert= alert
-    = yield
+    .container
+      %p
+        = stylesheet_link_tag 'https://stackpath.bootstrapcdn.com/bootstrap/4.5.2/css/bootstrap.min.css'
+        = stylesheet_link_tag 'application', media: 'all', 'data-turbolinks-track': 'reload'
+        = javascript_pack_tag 'application', 'data-turbolinks-track': 'reload'
```
Comment (danielpaul) [inline app/views/layouts/application.html.haml]: same with this

**Review by danielpaul — CHANGES_REQUESTED** [review]
Some comments for later for future branches: 
1. better naming for branch names. This might help to make it better: https://codingsight.com/git-branching-naming-convention-best-practices/
2. Better naming of the PR title.

File: config/routes.rb [hunk 5]
```diff
@@ -1,4 +1,5 @@
 Rails.application.routes.draw do
+  resources :companies
```
Comment (danielpaul) [inline config/routes.rb]: root usually comes first in the routes. So this line would be moved down under devise. The routes is a priority list of the most important ones at the top and lower ones coming down.

**Review by danielpaul — COMMENTED** [review]
I'm guessing that there are more comments to be pushed to this PR? Around Company model.

File: app/controllers/companies_controller.rb [hunk 6]
```diff
@@ -0,0 +1,75 @@
+class CompaniesController < ApplicationController
+  before_action :set_company, only: %i[show edit update destroy]
+  before_action :authenticate_user!, except: %i[index show]
```
Comment (danielpaul) [inline app/controllers/companies_controller.rb:2]: we would add this line to the start. So the controller runs this based on order. 

We don't have to `set_company` if the user is not logged in. So having the `authenticate_user!` check first would be best.

File: app/controllers/companies_controller.rb [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+  # before_action :authorized_user, only: [:edit, :update, :destroy]
```
Comment (danielpaul) [inline app/controllers/companies_controller.rb]: commented out code can be removed. we don't usually merge any commented code since we don't need them.

File: app/controllers/companies_controller.rb [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+  after_action :verify_authorized, except: %i[index]
+  after_action :verify_policy_scoped, only: %i[index]
+  # GET /companies or /companies.json
```
Comment (danielpaul) [inline app/controllers/companies_controller.rb]:
```suggestion
  
  # GET /companies or /companies.json
```

File: app/controllers/companies_controller.rb [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+  def index
+    @companies = policy_scope(Company)
+  end
+
+  # GET /companies/1 or /companies/1.json
+  def show
+    @companies = policy_scope(Company)
```
Comment (danielpaul) [inline app/controllers/companies_controller.rb]: do we need the list of companies at the show action?

File: app/controllers/companies_controller.rb [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+  end
```
Comment (danielpaul) [inline app/controllers/companies_controller.rb]: we should also be authorizing the action here. The policy can be hardcoded to be true since everyone can view this one.

File: app/controllers/companies_controller.rb [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+    authorize @company
+  end
+
+  # GET /companies/1/edit
+  def edit; end
+
+  # POST /companies or /companies.json
+  def create
+    @company = current_user.companies.create(company_params)
+    authorize @company
```
Comment (danielpaul) [inline app/controllers/companies_controller.rb:25]: so the autorizing line is happening after the line above which has used `.create`. Create would have already saved the data to our database? Need to verify this. If that is the case, then even when someone does not have permission, we would have saved their data. 

The solution would be to user `.new` which would not save. Then the code below where we save is where it actually saves which runs after someone has permission.

File: app/controllers/companies_controller.rb [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+    respond_to do |format|
+      if @company.save
+        format.html { redirect_to @company, notice: 'Company was successfully created.' }
+        format.json { render :show, status: :created, location: @company }
```
Comment (danielpaul) [inline app/controllers/companies_controller.rb]: We can remove json support. You can just remove the respond_to block - so it would just look like this:

```
if @company.save
    redirect_to @company, notice: 'Company was successfully created.'
else
    render :new, status: :unprocessable_entity
end
```

File: app/controllers/companies_controller.rb [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+      else
+        format.html { render :new, status: :unprocessable_entity }
+        format.json { render json: @company.errors, status: :unprocessable_entity }
+      end
+    end
+  end
+
+  # PATCH/PUT /companies/1 or /companies/1.json
+  def update
```
Comment (danielpaul) [inline app/controllers/companies_controller.rb:33]: need to authorise before updating. Otherwise anyone can update any records. 

Also same comment as above to remove json support for now.

File: app/controllers/companies_controller.rb [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+        format.json { render :show, status: :ok, location: @company }
+      else
+        format.html { render :edit, status: :unprocessable_entity }
+        format.json { render json: @company.errors, status: :unprocessable_entity }
+      end
+    end
+  end
+
+  # DELETE /companies/1 or /companies/1.json
+  def destroy
```
Comment (danielpaul) [inline app/controllers/companies_controller.rb:42]: need to authorise

File: app/controllers/companies_controller.rb [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+      format.json { head :no_content }
+    end
+  end
+
+  private
+
+  # Use callbacks to share common setup or constraints between actions.
+  def set_company
+    @company = Company.find(params[:id])
+    authorize @company
```
Comment (danielpaul) [inline app/controllers/companies_controller.rb]: aha... I see that you've done the authorising here. But we would do this in the action so it is explicitly mentioned in the actions. You can also control the position of the authorisation then.

File: app/controllers/companies_controller.rb (hunk 14 above)
Comment (danielpaul) [reply app/controllers/companies_controller.rb:42]: Ignore. Have commented below on `set_company` method.

File: app/controllers/companies_controller.rb (hunk 13 above)
Comment (danielpaul) [reply app/controllers/companies_controller.rb:33]: Ignore authorising comment. Have commented below on `set_company` method.

File: app/controllers/companies_controller.rb [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+  end
+
+  # Only allow a list of trusted parameters through.
+  def company_params
+    params.require(:company).permit(:name, :description, :user_id)
```
Comment (danielpaul) [inline app/controllers/companies_controller.rb]: we won't be allowing `user_id` here. That would mean that the browser can send us data for `user_id` and we would create the record for the user mentioned. So if I was logged in, I can post user_id of someone else and make DB record look like it was made by someone else to whom the record will be associated to.

Comment (danielpaul) [reply app/controllers/companies_controller.rb]: The user_id will be automatically set when we are creating it since we are going from current_user and going through it's relations.

File: app/helpers/companies_helper.rb [hunk 17]
```diff
@@ -0,0 +1,2 @@
+module CompaniesHelper
```
Comment (danielpaul) [inline app/helpers/companies_helper.rb]: We would usually remove empty generated files if we are not using them now. We can introduce these later if needed again.

Comment (danielpaul) [reply app/helpers/companies_helper.rb]: Same comment for the sass file above also.

File: app/models/user.rb [hunk 18]
```diff
@@ -3,7 +3,8 @@ class User < ApplicationRecord
   # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
   devise :database_authenticatable, :registerable, :confirmable,
          :recoverable, :rememberable, :validatable, :trackable
+  has_many :companies, dependent: :destroy
   validates :email, presence: true,
                     format: { with: /\A[^@\s]+@[^@\s]+\z/, message: 'must be a valid address' }
-  validates :first_name, :last_name, :password_confirmation, presence: true
+  validates :first_name, :last_name, presence: true
```
Comment (danielpaul) [inline app/models/user.rb:13]: let's try organise this file a little bit with comments and spacing so not all code is squished together. We would usually group all the validations together, relations, etc. And just above them add a empty line and add a comment saying the heading of the code below.

File: app/views/companies/_company.json.jbuilder [hunk 19]
```diff
@@ -0,0 +1,2 @@
+json.extract! company, :id, :name, :description, :created_at, :updated_at
```
Comment (danielpaul) [inline app/views/companies/_company.json.jbuilder]: we don't need this as we don't want to support json for now

File: app/views/companies/_form.html.haml [hunk 20]
```diff
@@ -0,0 +1,18 @@
...
+          %li= message
+
+  .field
+    = f.label :name
+    = f.text_field :name, class: "form-control"
+  .field
+    = f.label :description
+    = f.text_area :description, class: "form-control"
+  .field
+    = f.number_field :user_id, id: :company_user_id, class: "form-control", value: current_user.id, type: :hidden
```
Comment (danielpaul) [inline app/views/companies/_form.html.haml]: don't need this one. We don't want this to be posted from the client side. We will use the current_user in the background.

File: app/views/companies/index.json.jbuilder [hunk 21]
```diff
@@ -0,0 +1 @@
+json.array! @companies, partial: "companies/company", as: :company
```
Comment (danielpaul) [inline app/views/companies/index.json.jbuilder]: file can be deletd

File: app/views/companies/show.json.jbuilder [hunk 22]
```diff
@@ -0,0 +1 @@
+json.partial! "companies/company", company: @company
```
Comment (danielpaul) [inline app/views/companies/show.json.jbuilder]: file can be deleted

File: config/webpacker.yml [hunk 23]
```diff
@@ -83,7 +83,7 @@ production:
   <<: *default
 
   # Production depends on precompilation of packs prior to booting for performance.
-  compile: false
+  compile: true
```
Comment (danielpaul) [inline config/webpacker.yml]: Why is this needed to be `true`?

File: db/migrate/20210411224620_add_user_id_to_companies.rb [hunk 24]
```diff
@@ -0,0 +1,6 @@
+class AddUserIdToCompanies < ActiveRecord::Migration[6.1]
+  def change
+    add_column :companies, :user_id, :integer
+    add_index :companies, :user_id
```
Comment (danielpaul) [inline db/migrate/20210411224620_add_user_id_to_companies.rb]: have a look at these ones:

```
add_reference :companies, :user, index: true
add_foreign_key :companies, :users
```

This would do a lot of other things for us.

Comment (danielpaul) [reply db/migrate/20210411224620_add_user_id_to_companies.rb]: You can look up these method names to see the difference between these ones and add_column + add_index

File: spec/requests/companies_spec.rb [hunk 25]
```diff
@@ -0,0 +1,11 @@
...
+
+RSpec.describe '/companies', type: :request do
+  describe 'GET /index' do
+    it 'renders a successful response' do
+      get companies_path
+
+      expect(response).to have_http_status(:success)
+    end
+  end
+end
```
Comment (danielpaul) [inline spec/requests/companies_spec.rb:23]: other tests to add:

- [x] * user trying to access new company page but not logged in -> should redirect them to the login page
- [x] + similar tests for other pages that require user to be logged in.

- [x] * logged in as User 1 and trying to edit User 1's company -> allowed.
- [x] * logged in as User 1 and trying to edit User 2's company -> not allowed. 
- [x] + Similar tests as above for update, destroy actions. 

- [x] * correct success responses from all actions

**Review by danielpaul — CHANGES_REQUESTED** [review]

File: config/webpacker.yml (hunk 23 above)
Comment (jacobperia) [reply config/webpacker.yml]: had some issues with webpacker and i was trying to fix it for a long time and one of the forums told me to change it and i forgot to change back.

File: app/controllers/companies_controller.rb (hunk 16 above)
Comment (danielpaul) [reply app/controllers/companies_controller.rb]: This looks like it still need to be updated. To remove user_id from this list.

**Review by danielpaul — CHANGES_REQUESTED** [review]

**Review by danielpaul — APPROVED** [review]
